### ISSUE TITLE
custom openssl issue -> segmentation fault (centos 7)

### ISSUE THREAD

**Issue by cc-brian** [kaltura/nginx-vod-module#563]
This issue is the same as this closed issue:

https://github.com/kaltura/nginx-vod-module/issues/411

However, I do need DRM/HLS-AES supported.

Basically, installing nginx 1.11.10 with openssl 1.0.2k to support HTTP2 on Centos 7 results in a segmentation fault with nginx-vod-module added.  If I remove it, everything works fine.

How do I install nginx-vod-module referencing  a custom openssl source?

Here's my command line:

./configure --user=nginx --group=nginx --prefix=/etc/nginx --sbin-path=/usr/sbin/nginx --conf-path=/etc/nginx/nginx.conf --error-log-path=/var/log/nginx/error.log --http-log-path=/var/log/nginx/access.log --pid-path=/var/run/nginx.pid --lock-path=/var/run/nginx.lock --with-threads --with-file-aio --with-http_ssl_module --with-pcre --with-http_v2_module --with-http_sub_module  --with-http_auth_request_module  --with-http_stub_status_module --with-openssl=$HOME/sources/openssl-1.0.2k --add-module=$HOME/sources/nginx-vod-module-master
make
make install

Please advise.

Thanks!

**Comment by erankor** [comment]
export C_INCLUDE_PATH & LIBRARY_PATH to point to the openssl include & lib folders, so that this custom openssl will be found using the search path and not using with-openssl

**Comment by cc-brian** [comment]
Thanks for the super fast reply!

So if I understand you correctly on the command line I would type:

$ C_INCLUDE_PATH=$HOME/sources/openssl-1.0.2k/include
$ export C_INCLUDE_PATH

$ LIBRARY_PATH=$HOME/sources/openssl-1.0.2k/(**not sure which folder is library??**)
$ export LIBRARY_PATH

Questions:

1) Does this affect only the current shell or the whole OS?  
2) To avoid any issues with the OS going forward should I change it back to default after install?  If so, how do I get it back to default?

Sorry for the basic questions, never did this before.

Thanks again

**Comment by erankor** [comment]
The library folder will not be under the source code, it will probably be something like /usr/local/ssl/lib/.
1. It affects the current shell and any processes it creates
2. Best way would be to echo these vars before you change them, on the environments I use, they are not set to anything, I have a script under /etc/profile.d/ that sets them to what I need.

**Comment by jessp01** [comment]
Also note that for runtime loading of the OpenSSL shared objects [assuming you're not statically linking against the archive files], you should export LD_LIBRARY_PATH.

**Comment by cc-brian** [comment]
Thanks for all the info, but I'm still a little foggy here.

I understand the basic idea of setting env variables realtime in the shell now, thanks.  However, I'm not sure about the ssl libraries.

For example, this part makes sense, as I'm using the newly downloaded ssl source.

C_INCLUDE_PATH=$HOME/sources/openssl-1.0.2k/include
export $C_INCLUDE_PATH

However, you mentioned the library path would not be in the new source files, but somewhere on the system.  In my case, the only files I've downloaded are the openssl source files from openssl.org.  So why would I set my library path to the older installed files on my system which are a different and older version?

As a reminder, to compile nginx I simply download the source files for openssl and reference them in my configure syntax, then make, make install as mentioned above.  I don't actually install a new version of openssl anywhere on the system.

Any guidance is much appreciated!

**Comment by jessp01** [comment]
Hi @cc-brian,

This really isn't specific to the VOD module at all but in general:
When linking against third party libs in C, you can either link against them statically, in which case, you will need the archive files [.a] or dynamically, in which case you'll need the shared object files [SO]. Either way, you will need the header files [.h].
When linking statically, there is no need to export LD_LIBRARY_PATH since the openSSL archives will be linked into the nginx binary, when using dynamic linkage, it is needed for runtime loading of the libs.

For example, on my machine where the kaltura-nginx deb package is installed:
\# ldd /opt/kaltura/nginx/sbin/nginx |grep ssl
        libssl.so.1.0.0 => /lib/x86_64-linux-gnu/libssl.so.1.0.0 (0x00007f41b2f24000)
since it is under /lib, there is no need to export LD_LIBRARY_PATH, if it were under say /my/own/ssl/path, then it would have been necessary because LD needs to know where to look.
LIBRARY_PATH is needed during the compilation process and needs to point to where your openSSL archive files/shared libs reside.

Bottom line: you MUST first compile openSSL from the source you've downloaded. The prefix for the target is determined by the --prefix flag passed to the configure script. If --prefix is not passed, it will resort to /usr/local, so, libs and archives will go to /usr/local/lib and header files will go to /usr/local/include
